drivers/sdhc/sdhc_stm32.c

```diff
@@ -0,0 +1,659 @@
+						      (uint32_t *)&cmd->response);
+		break;
+
+	case SDIO_RW_DIRECT:
+		res = sdhc_stm32_rw_direct(dev_data, cmd);
+		break;
+
+	case SDIO_RW_EXTENDED:
+		res = sdhc_stm32_rw_extended(dev_data, cmd, data);
+		break;
+
+	case SD_VOL_SWITCH:
+		res = sdhc_stm32_switch_to_1_8v(dev);
+		break;
+
+	default:
+		res = -ENOTSUP;
+		LOG_DBG("Unsupported Command, opcode:%d\n.", cmd->opcode);
+		break;
+	}
+
+	if (res != HAL_OK) {
+		LOG_ERR("Command Failed, opcode:%d\n.", cmd->opcode);
+		sdhc_stm32_log_err_type(&dev_data->hsd);
+	}
+
+	k_mutex_unlock(&dev_data->bus_mutex);
+	(void)pm_device_runtime_put(dev);
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	return res;
+}
+
+static int sdhc_stm32_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	struct sdhc_stm32_data *data = dev->data;
+	struct sdhc_io *host_io = &data->host_io;
+	int res = 0;
+
+	(void)pm_device_runtime_get(dev);
+	/* Prevent the clocks to be stopped during the request */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	k_mutex_lock(&data->bus_mutex, K_FOREVER);
+
+	if ((ios->clock != 0) && (host_io->clock != ios->clock)) {
+		if ((ios->clock > STM32_SDIO_F_MAX) || (ios->clock < STM32_SDIO_F_MIN)) {
+			LOG_ERR("Invalid Clock Frequency\n");
+			return -EINVAL;
+		}
+		res = HAL_SDIO_ConfigFrequency(&data->hsd, (uint32_t)ios->clock);
+		if (res != HAL_OK) {
+			LOG_ERR("Failed to set clock as %d\n", ios->clock);
+		} else {
+			host_io->clock = ios->clock;
+			LOG_INF("set SDHC clock as %d\n", ios->clock);
+		}
+		k_msleep(10);
+	}
+
+	if (ios->power_mode == SDHC_POWER_OFF) {
+		(void)SDMMC_PowerState_OFF(data->hsd.Instance);
+		k_msleep(10);
+	} else if (ios->power_mode == SDHC_POWER_ON) {
+		(void)SDMMC_PowerState_ON(data->hsd.Instance);
+		k_msleep(10);
+	}
+
+	if ((ios->bus_width != 0) && (host_io->bus_width != ios->bus_width)) {
+		MODIFY_REG(data->hsd.Instance->CLKCR, SDMMC_CLKCR_WIDBUS,
+			   (ios->bus_width == SDHC_BUS_WIDTH4BIT) ? SDMMC_BUS_WIDE_4B
+								  : SDMMC_BUS_WIDE_1B);
+		host_io->bus_width = ios->bus_width;
+	}
+
+	k_mutex_unlock(&data->bus_mutex);
+	(void)pm_device_runtime_put(dev);
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	return res;
+}
+
+static void sdhc_stm32_init_props(const struct device *dev)
+{
+	const struct sdhc_stm32_config *sdhc_config = (struct sdhc_stm32_config *)dev->config;
+	struct sdhc_stm32_data *data = dev->data;
+	struct sdhc_host_props *props = &data->props;
+
+	memset(props, 0, sizeof(struct sdhc_host_props));
+
+	props->f_min = SD_CLOCK_25MHZ;
+	props->f_max = SD_CLOCK_208MHZ;
+	props->power_delay = sdhc_config->power_delay_ms;
+	props->host_caps.vol_330_support = true;
+
+#ifdef CONFIG_SDHC_1_8V_SUPPORT
+	props->host_caps.vol_180_support = true;
+#else
+	props->host_caps.vol_180_support = false;
+#endif
+	props->host_caps.vol_300_support = false;
+	props->host_caps.bus_8_bit_support = false;
```

Comment:
I suspect at least one of these properties is true for the IP, likely both- glancing at the SDMMC IP on the STM32H742, it appears to support up to HS200, and 4/8 bit bus modes.

If these properties vary between IP instantiations, we should make them configurable in DT. We should not outright disable them in DT though, because that will prevent this driver from working well with the SD or eMMC stack.